File: python/bob/bob.py
'''
Implemented Bob's phrases as a dictionary this time to make it easier to
change, add or delete his phrases in the future. This could be stored in a
separate file, for instance.
'''

bobs_phrases = {    "nothing":            'Fine. Be that way!',
                    "forceful_question":  'Calm down, I know what I\'m doing!',
                    "shout":              'Whoa, chill out!',
                    "question":           'Sure.',
                    "general":            'Whatever.'
                
                }


def hey(phrase):
    '''
    Bob responds to you based on what you say
    
    Inputs: phrase (string): whatever you want to say to Bob
    
    Returns: Bob's response to you (string)
    
    '''
    
    # Altered logic to simplify code: strip phrase in advance of if clauses
    # and nest the two conditions that depend on UPPER CASE
    #                   Credit for idea to: zaneundrscr
    
    phrase = phrase.strip()
    
    # if you say nothing
    if not phrase:
        return bobs_phrases["nothing"]
    
    elif phrase.isupper():
        # if you shout [CAPS] and ask a question [? at end]
        if phrase[-1] == '?':
            return bobs_phrases["forceful_question"]
    
        # if you just shout [CAPS]
        else:
            return bobs_phrases["shout"]
    
    # if you just ask a question [? at end]
    elif phrase[-1] == '?':
        return bobs_phrases["question"]
     
    # all other cases
    else:
        return bobs_phrases["general"]

File: haskell/bob/src/Bob.hs
{-# LANGUAGE OverloadedStrings #-}

module Bob (responseFor) where

import           Data.Char (isLower, isSpace, isUpper)
import           Data.Text (Text)
import qualified Data.Text as T

questionResponse, yellingResponse, yellingQuestionResponse, silenceReponse, defaultResponse :: Text
questionResponse        = "Sure."
yellingResponse         = "Whoa, chill out!"
yellingQuestionResponse = "Calm down, I know what I'm doing!"
silenceReponse          = "Fine. Be that way!"
defaultResponse         = "Whatever."

responseFor :: Text -> Text
responseFor t
    | yelling && question   = yellingQuestionResponse
    | yelling               = yellingResponse
    | question              = questionResponse
    | isSilence t           = silenceReponse
    | otherwise             = defaultResponse
    where yelling  = isYelling t
          question = isQuestion t

isQuestion :: Text -> Bool
isQuestion t = not (T.null strippedText) 
            && T.last strippedText == '?'
    where strippedText = T.stripEnd t

isYelling :: Text -> Bool
isYelling t = not (T.any isLower t) 
           && T.any isUpper t

isSilence :: Text -> Bool
isSilence = T.all isSpace
